docs/architecture/adr-012-handlers-redesign.md

```diff
@@ -0,0 +1,598 @@
+# ADR 012: Handlers validation and execution separation
+
+## Status
+Proposed
+
+## Changelog
+* 9/9/2022: initial proposal
+
+## Context
+
+The main purpose of this ADR is to split each IBC handler (e.g. `UpdateClient`, `ConnOpenTry`, etc.) into a "validation" and an "execution" phase in order to accomodate a larger class of host architectures such as Namada. We call "validation" the part of a handler that does all the checks necessary for correctness. We call "execution" the part of the handler that applies the state modifications, assuming that validation passed. 
+
+Our current `deliver()` entrypoint can then be split into 2 entrypoints: `validate()` and `execute()`. More specifically, we replace the current `Ics26Context` and all its supertraits with 2 traits: `ValidationContext` and `ExecutionContext`. `ValidationContext` exposes and implements the `validate()` entrypoint, while the `ExecutionContext` exposes and implements the `execute()` entrypoint.
```

Comment:
We will also need a `checkTx` equivalent which might not be the same as `validate()`. https://github.com/informalsystems/ibc-rs/issues/2094